### PR TITLE
#3058 New delay implementation working with IDA.

### PR DESCRIPTION
**Pull request by gautierbureau**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] the corresponding milestone was added in the ticket and in this PR
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated

### CODE REVIEW

File: dynawo/sources/Common/Dictionaries/DYNLog_en_GB.dic [hunk 1]
```diff
@@ -204,8 +204,9 @@ CalculateICIteration          =             initial condition iteration %1%
 IncoherentParamMinimumModeChangeType =      parameter minimumModeChangeTypeForAlgebraicRestoration (value=%1%) of solver should be one of DIFFERENTIAL, ALGEBRAIC or ALGEBRAIC_J_UPDATE. Default value will be used (ALGEBRAIC).
 // --> DYNSolverIDA
 SolverIDAInitOk               =             initialization of IDA solver : ok
-SolverIDABeforeCalcIC         =             BEFORE CALC IC
-SolverIDAAfterInit            =             AFTER INIT
+SolverIDAStartCalculateIC     =             start CalculateIC
+SolverIDABeforeCalcIC         =             before CalculateIC
+SolverIDAAfterInit            =             after init
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Common/Dictionaries/DYNLog_en_GB.dic:209]: The logs already existed but changed a bit its output to be more useful

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp [hunk 2]
```diff
@@ -106,7 +113,7 @@ DelayManager::loadDelays(const std::vector<std::string>& values) {
 
       if (!last_time) {
         last_time = value;
-      } else if (doubleEquals(time, *last_time) || *last_time > time) {
+      } else if (*last_time > time) {
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:116]: With IDA and the precision we dump the saved time points we can have something like this `;5,5.5;5,5.5;`, I figured that I would not be an issue to restart a simulation with this points and to the only case we need an error is when last_time > time ie we would have something like `;5.1,5.5;5,5.5;`. Maybe I am wrong...

File: dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp [hunk 3]
```diff
@@ -838,6 +839,9 @@ ModelManager::loadParameters(const string& parameters) {
     throw DYNError(Error::MODELER, WrongDataNum, parametersFileName().c_str());
   }
 
+  // To activate all delays
+  delayManager_.evalMode(getCurrentTime());
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:843]: When we restart the simulation from a dump we need to activate all delays before we start the simulation

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_MacroConnects/reference/outputs/initValues/globalInit/dumpInitValues-GEN____3_SM.txt [hunk 4]
```diff
@@ -1,9 +1,9 @@
- ====== INIT VARIABLES VALUES ======
-generator_lambdaDPu                               : y =       1.015355 yp =       0.000010
-generator_lambdaQ1Pu                              : y =      -0.000002 yp =       0.000030
-generator_lambdaQ2Pu                              : y =      -0.000002 yp =       0.000163
-generator_lambdafPu                               : y =       1.137650 yp =      -0.000001
-generator_omegaPu_value                           : y =       1.000000 yp =       0.000004
+ ====== VARIABLES VALUES ======
+generator_lambdaDPu                               : y =       1.015355 yp =       0.000000
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_MacroConnects/reference/outputs/initValues/globalInit/dumpInitValues-GEN____3_SM.txt:2]: I changed `int flag = IDACalcIC(IDAMem_, IDA_YA_YDP_INIT, initStep_);` before 10. was hardcoded and I don't know from where this value came. And when we restart the simulation from IDA it enables to restart with the same time step as the previous one

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:116]: Maybe we could adapt the precision of the dump to the precision of the simulation as this might be misleading

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp [hunk 5]
```diff
@@ -152,11 +160,22 @@ DelayManager::setGomc(state_g* const p_glocal, size_t offset) {
   }
 }
 
+void
+DelayManager::evalMode(const double time) {
+  boost::unordered_map<size_t, Delay>::iterator it;
+  for (it = delays_.begin(); it != delays_.end(); ++it) {
+    double delayTime = it->second.getDelayMax();
+    if (!(time < delayTime || doubleEquals(time, delayTime))) {
+      it->second.trigger();
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:169]: Yes it seems I need to test it

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:116]: I think with IDA we would even need to dump with machine precision

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 6]
```diff
@@ -813,8 +813,10 @@ Simulation::init() {
 
   if (Trace::logExists(Trace::modeler(), DEBUG))
     model_->printModel();
-  if (Trace::logExists(Trace::variables(), DEBUG))
+  if (Trace::logExists(Trace::variables(), DEBUG)) {
+    model_->evalDynamicYType();
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Simulation/DYNSimulation.cpp:817]: I think this modification was not supposed to go here I will remove it

[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Simulation/DYNSimulation.cpp:817]: It was supposed to go with the type added to the debug log file, the issue was that the type is not initialized at the moment of log creation. But everything seemed ok in the log

File: dynawo/sources/Modeler/Common/test/TestDelay.cpp [hunk 7]
```diff
@@ -297,8 +298,8 @@ compare(const std::vector<std::string>& lhs, const std::vector<std::string>& rhs
 }
 
 TEST(CommonTest, testDelayManagerClassParameters) {
-  std::string formatted1 = "10:1,1.1;2,2.2;3,3.3;4,4.4;5,5.5;";
-  std::string formatted2 = "20:1,1.1;2,2.2;3,3.3;4,4.4;5,5.5;";
+  std::string formatted1 = "10:1.5:1,1.1;2,2.2;3,3.3;4,4.4;5,5.5;";
+  std::string formatted2 = "20:3:1,1.1;2,2.2;3,3.3;4,4.4;5,5.5;";
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/test/TestDelay.cpp:302]: evalG is tested in the function but evalMode is not called as `manager.triggerDelay(id2);` is the "same" call it would do. I will modify the test to call evalMode

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:116]: And then in the RingBuffer the value is ignored

``` cpp
if (queue_.size() > 0 && doubleEquals(queue_.back().first, time)) {
    // ignore if we add multiple time the same value
    return;
  }
```

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Simulation/DYNSimulation.cpp:817]: the dynamic types might use variables values to determine which type is used. At this point at the dump was not loaded (I d'ont know for the getY0) we might get a crash

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:116]: we need at least dynawo precision to make sure this doubleEquals is correct

File: dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:843]: only if the time is longer than the startTime right? (I didn't check if it was the case in the current implementation)

[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:843]: If the start time is lower it will simply doesn't activate the delay

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:132]: I updated the implementation

[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:132]: if I'm not mistaken doubleEquals(time, *last_time) should not happen as we register after the timestep is validated?

[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.cpp:132]: If fact it should not happen now that we dump with dynawo precision et the test mentioned before so I remove it and go back to the initial code on this part

File: dynawo/sources/Modeler/Common/DYNDelayManager.cpp [hunk 8]
```diff
@@ -107,7 +114,7 @@ DelayManager::loadDelays(const std::vector<std::string>& values) {
       if (!last_time) {
         last_time = value;
       } else if (doubleEquals(time, *last_time) || *last_time > time) {
-        // last_time >= time : error
+        // last_time > time : error
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDelayManager.cpp:117]: the comment should be reverted back too

File: dynawo/sources/Modeler/Common/DYNDelay.cpp [hunk 9]
```diff
@@ -23,12 +23,24 @@
...
-Delay::Delay(const std::vector<std::pair<double, double> >& timepoints) : time_(NULL), value_(NULL), buffer_(0), initialValue_(), trigger_() {
+Delay::Delay(const std::vector<std::pair<double, double> >& timepoints, double delayMax, double time) :
+time_(NULL),
+value_(NULL),
+buffer_(delayMax),
+delayActivated_(false) {
   for (std::vector<std::pair<double, double> >::const_iterator it = timepoints.begin(); it != timepoints.end(); ++it) {
     buffer_.addNoCheck(it->first, it->second);
   }
+  delayMax_ = delayMax;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDelay.cpp:42]: should be done in init list?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

**Comment by gautierbureau** [comment]
This version should be ok !

File: dynawo/sources/Modeler/Common/DYNDelay.h [hunk 10]
```diff
@@ -112,38 +115,55 @@ class Delay {
...
+    return delayTime_;
+  }
+
+  /**
+   * @brief Set the time of the delay
+   *
+   * @param delayTime the time of the delay
+   */
+  void setDelayTime(double delayTime) {
+    if (!doubleEquals(delayTime, delayTime_))
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDelay.h:145]: Why is this needed?

File: dynawo/sources/Modeler/Common/DYNDelayManager.h [hunk 11]
```diff
@@ -144,28 +146,31 @@ class DelayManager {
...
 
+  /**
+  * @brief Retrieves the delay by id (non-const version)
+  *
+  * Precondition: the id is acceptable
+  *
+  * @param id the id to use
+  * @returns the corresponding delay
+  */
+  Delay& getNonCstDelayById(size_t id) const {
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDelayManager.h:168]: There should be a setter in Delay manager

File: dynawo/sources/Modeler/Common/DYNRingBuffer.cpp [hunk 12]
```diff
@@ -87,6 +87,7 @@ RingBuffer::get(double time, double delay) const {
     // case first element: it would mean that delay is greater than max delay
     // shouldn't happen by construction with function add
     throw DYNError(DYN::Error::SIMULATION, IncorrectDelay, delay, time, maxDelay_);
+    // return found->second;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNRingBuffer.cpp:90]: please remove this useless line

File: dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp [hunk 13]
```diff
@@ -1435,21 +1434,17 @@ ModelManager::computeDelay(int exprNumber, double exprValue, double time, double
...
-  if (time < delayTime || doubleEquals(time, delayTime)) {
-    // requested time is allowed but the simulation has not run for enought time to compute the delayed value:
-    // use the initial value to avoid artefacts
-    const boost::optional<double>& initialValue = delayManager_.getInitialValue(exprNumber);
-#if _DEBUG_
-    // shouldn't happen by construction
-    assert(initialValue.is_initialized());
-#endif
+  const Delay& delay = delayManager_.getDelayById(exprNumber);
+  delayManager_.getNonCstDelayById(exprNumber).setDelayTime(delayTime);
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:1438]: I really don't get why this is necessary

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/Common/DYNDelayManager.h (hunk 11 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNDelayManager.h:168]: You mean a setDelayTime ?

File: dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp (hunk 13 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:1438]: To handle a change in the delay time parameter

File: dynawo/sources/Modeler/Common/DYNRingBuffer.cpp (hunk 12 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/Common/DYNRingBuffer.cpp]: done

File: dynawo/sources/Modeler/Common/DYNDelayManager.h (hunk 11 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNDelayManager.h:168]: yes

File: dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp (hunk 13 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:1438]: I guess we will need to discuss so that I can understand. Anyway the computeDelay is probably not a good place to modify the delay

Comment (gautierbureau) [reply dynawo/sources/Modeler/ModelManager/DYNModelManager.cpp:1438]: This is the only moment where delayTime is updated from the Modelica model when it changes like in of the nrt

**Review by rosiereflo — APPROVED** [review]
